LUCENE-10315: Speed up BKD leaf block ids codec by a 512 ints ForUtil

gf2121 opened the pull request:
Elasticsearch (which based on lucene) can automatically infers types for users with its dynamic mapping feature. When users index some low cardinality fields, such as gender / age / status... they often use some numbers to represent the values, while ES will infer these fields as long, and ES uses BKD as the index of long fields. When the data volume grows, building the result set of low-cardinality fields will make the CPU usage and load very high.

This is a flame graph we obtained from the production environment:
[flame.svg](https://issues.apache.org/jira/secure/attachment/13037560/13037560_addall.svg)

It can be seen that almost all CPU is used in addAll. When we reindex long to keyword, the cluster load and search latency are greatly reduced ( We spent weeks of time to reindex all indices... ). I know that ES recommended to use keyword for term/terms query and long for range query in the document, but there are always some users who didn't realize this and keep their habit of using sql database, or dynamic mapping automatically selects the type for them. All in all, users won't realize that there would be such a big difference in performance between long and keyword fields in low cardinality fields. So from my point of view it will make sense if we can make BKD works better for the low/medium cardinality fields.

As far as i can see, for low cardinality fields, there are two advantages of keyword over long:

1. ForUtil used in keyword postings is much more efficient than BKD's delta VInt, because its batch reading (readLongs) and SIMD decode.
2. When the query term count is less than 16, TermsInSetQuery can lazily materialize of its result set, and when another small result clause intersects with this low cardinality condition, the low cardinality field can avoid reading all docIds into memory.

This approach is targeting to solve the first point. The basic idea is trying to use a 512 ints ForUtil for BKD ids codec. I benchmarked this optimization by mocking some random LongPoint and querying them with PointInSetQuery.

**Benchmark Result**
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/gf/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/gf/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">

<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
.font5
	{color:windowtext;
	font-size:9.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:等线;
	mso-generic-font-family:auto;
	mso-font-charset:134;}
tr
	{mso-height-source:auto;
	mso-ruby-visibility:none;}
col
	{mso-width-source:auto;
	mso-ruby-visibility:none;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:12.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:等线;
	mso-generic-font-family:auto;
	mso-font-charset:134;
	mso-number-format:General;
	text-align:general;
	vertical-align:middle;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
.xl65
	{color:#172B4D;
	font-size:14.0pt;}
.xl66
	{color:#172B4D;
	font-size:14.0pt;
	mso-number-format:Percent;}
.xl67
	{font-size:14.0pt;}
ruby
	{ruby-align:left;}
rt
	{color:windowtext;
	font-size:9.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:等线;
	mso-generic-font-family:auto;
	mso-font-charset:134;
	mso-char-type:none;
	display:none;}
-->

</head>

<body link="#0563C1" vlink="#954F72">
<meta charset=utf-8>



doc count | field   cardinality | query   point | baseline   QPS | candidate   QPS | diff   percentage
-- | -- | -- | -- | -- | --
100000000 | 32 | 1 | 51.44 | 148.26 | 188.22%
100000000 | 32 | 2 | 26.8 | 101.88 | 280.15%
100000000 | 32 | 4 | 14.04 | 53.52 | 281.20%
100000000 | 32 | 8 | 7.04 | 28.54 | 305.40%
100000000 | 32 | 16 | 3.54 | 14.61 | 312.71%
100000000 | 128 | 1 | 110.56 | 350.26 | 216.81%
100000000 | 128 | 8 | 16.6 | 89.81 | 441.02%
100000000 | 128 | 16 | 8.45 | 48.07 | 468.88%
100000000 | 128 | 32 | 4.2 | 25.35 | 503.57%
100000000 | 128 | 64 | 2.13 | 13.02 | 511.27%
100000000 | 1024 | 1 | 536.19 | 843.88 | 57.38%
100000000 | 1024 | 8 | 109.71 | 251.89 | 129.60%
100000000 | 1024 | 32 | 33.24 | 104.11 | 213.21%
100000000 | 1024 | 128 | 8.87 | 30.47 | 243.52%
100000000 | 1024 | 512 | 2.24 | 8.3 | 270.54%
100000000 | 8192 | 1 | 3333.33 | 5000 | 50.00%
100000000 | 8192 | 32 | 139.47 | 214.59 | 53.86%
100000000 | 8192 | 128 | 54.59 | 109.23 | 100.09%
100000000 | 8192 | 512 | 15.61 | 36.15 | 131.58%
100000000 | 8192 | 2048 | 4.11 | 11.14 | 171.05%
100000000 | 1048576 | 1 | 2597.4 | 3030.3 | 16.67%
100000000 | 1048576 | 32 | 314.96 | 371.75 | 18.03%
100000000 | 1048576 | 128 | 99.7 | 116.28 | 16.63%
100000000 | 1048576 | 512 | 30.5 | 37.15 | 21.80%
100000000 | 1048576 | 2048 | 10.38 | 12.3 | 18.50%
100000000 | 8388608 | 1 | 2564.1 | 3174.6 | 23.81%
100000000 | 8388608 | 32 | 196.27 | 238.95 | 21.75%
100000000 | 8388608 | 128 | 55.36 | 68.03 | 22.89%
100000000 | 8388608 | 512 | 15.58 | 19.24 | 23.49%
100000000 | 8388608 | 2048 | 4.56 | 5.71 | 25.22%



</body>

</html>


The indices size is reduced for low cardinality fields and flat for high cardinality fields.

```
113M    index_100000000_doc_32_cardinality_baseline
114M    index_100000000_doc_32_cardinality_candidate

140M    index_100000000_doc_128_cardinality_baseline
133M    index_100000000_doc_128_cardinality_candidate

241M    index_100000000_doc_8192_cardinality_baseline
233M    index_100000000_doc_8192_cardinality_candidate

193M    index_100000000_doc_1024_cardinality_baseline
174M    index_100000000_doc_1024_cardinality_candidate

314M    index_100000000_doc_1048576_cardinality_baseline
315M    index_100000000_doc_1048576_cardinality_candidate

392M    index_100000000_doc_8388608_cardinality_baseline
391M    index_100000000_doc_8388608_cardinality_candidate
```

gf2121 committed:
stash

gf2121 committed:
stash

gf2121 committed:
check

gf2121 committed:
add forutil

gf2121 committed:
name codes

gf2121 committed:
note

gf2121 committed:
for util

gf2121 committed:
reduce code num

gf2121 committed:
spotless

jpountz commented:
Nice. I wonder if we need to specialize for so many numbers of bits per value like we do for postings, or if we should only specialize for a few numbers of bits per value that are both useful and fast, e.g. 0, 4, 8, 16, 24 and 32.

gf2121 commented:
Thanks @jpountz ! I have two points to conifrm:

1. Do you mean only **specialize** a few numbers of bits per value (other bpvs using `decodeslow`), or just **support** these numbers when writing (round up bpvs like what we did in DirectWriter) ?

2. This issue is having the same problem as https://github.com/apache/lucene/pull/545 that constants with complex names making codes harder to read, maybe we need to solve it first? I made some change in that approach, hoping it will make sense to you :)

jpountz commented:
I was thinking of only supporting these numbers of bits per value indeed. For postings, numbers are always deltas, so we can generally expect them to be small. But for BKD trees it tends to be more an exception so I don't think we should spend too much effort on supporting so many bits per value and only focus on the ones that matter:
 - 32 bits per value for large segments where the doc ID order is random.
 - 24 bits per value for medium segments (less than 2^24 docs) where the doc ID order is random.
 - 16 bits per value plus delta coding from the minimum doc ID in the block for the case where there is some clustering of doc IDs.
 - And maybe the bitset strategy you added recently already covers the other cases like sorted indexes and values that exist in many docs, so that we don't need the delta-coding between consecutive anymore, which is slow anyway due to the cumulative sum?

gf2121 committed:
limit bpv to 16/24/32 and using floor delta codec

gf2121 committed:
make diff a bit more beautiful

gf2121 committed:
iter

gf2121 committed:
assert count

gf2121 committed:
make writer final

gf2121 committed:
iter

gf2121 commented:
Thanks @jpountz for the great advice! I have implemented it and codes are simplified a lot.

gf2121 committed:
try to make remainder also SIMD

gf2121 commented on lucene/core/src/java/org/apache/lucene/util/bkd/BKDForUtil.java:
```diff
@@ -0,0 +1,126 @@
+    for (int i = 0; i < 192; ++i) {
+      tmp[i] = longs[i] << 8;
+    }
+    for (int i = 0; i < 64; i++) {
```
This encoding logic is a bit different from the `ForUtil` we used in postings, as i want to make the remainder decoding can also trigger SIMD. Here is the JMH result:

```
Benchmark                            (bitsPerValue)  (byteOrder)   Mode  Cnt        Score         Error  Units
PackedIntsDecodeBenchmark.baseline               24           LE  thrpt    5  7756530.597 ± 1654468.198  ops/s
PackedIntsDecodeBenchmark.candidate              24           LE  thrpt    5  9681438.494 ± 2528482.525  ops/s

```

gf2121 committed:
plus when expand

gf2121 commented on lucene/core/src/java/org/apache/lucene/util/bkd/BKDForUtil.java:
```diff
@@ -0,0 +1,126 @@
+    return expandMask32((1L << bitsPerValue) - 1);
+  }
+
+  private static void expand16(long[] arr, final long base) {
```
This is a bit tricky but indeed helping the performance:

```
Benchmark                                  (bitsPerValue)  (byteOrder)   Mode  Cnt         Score        Error  Units
PackedIntsDecodeBenchmark.plusAfterExpand              16           LE  thrpt   20   7996083.715 ± 618197.203  ops/s
PackedIntsDecodeBenchmark.plusWhenExpand               16           LE  thrpt   20  10681376.808 ± 542945.909  ops/s

```

jpountz replied:
Thanks for creating these micro benchmarks.

gf2121 committed:
judge cluster should not rely on sorted

gf2121 committed:
add an assert

gf2121 committed:
assert is making CI angry, remove

jpountz commented on lucene/core/src/java/org/apache/lucene/util/bkd/BKDForUtil.java:
```diff
@@ -0,0 +1,126 @@
+// This file has been automatically generated, DO NOT EDIT
```
It no longer looks auto-generated? (which sounds good to me since it's much simpler now)

jpountz reviewed (commented):
Thanks @gf2121.

If you have the time, I'd be curious to see how it would compare with a similar approach where we would use an int[] to hold the doc IDs instead of longs. Postings started with longs because it helped do "fake" SIMD operations, e.g. summing up a single long would sum up two ints under the hood. But as our BKD tree doesn't need to perform prefix sums, it doesn't need such tricks. We could add a new `IndexInput#readInts` similar to `IndexInput#readLongs` and see how the impl that uses longs compares to the one that uses ints?

gf2121 commented:
Thanks @jpountz !

> I'd be curious to see how it would compare with a similar approach where we would use an int[] to hold the doc IDs instead of longs.

I wrote a simple micro benchmark and result looks promising (especially for 32 bit) , I'd like to give it a try.

```
Benchmark                                 (byteOrder)   Mode  Cnt         Score         Error  Units
PackedIntsDecodeBenchmark.decode16ToInt            LE  thrpt   10  15603376.770 ±  229121.911  ops/s
PackedIntsDecodeBenchmark.decode16ToLong           LE  thrpt   10  11128632.044 ± 1037046.154  ops/s
PackedIntsDecodeBenchmark.decode24ToInt            LE  thrpt   10   9494705.748 ± 1773057.040  ops/s
PackedIntsDecodeBenchmark.decode24ToLong           LE  thrpt   10   7741109.174 ±  838010.286  ops/s
PackedIntsDecodeBenchmark.decode32ToInt            LE  thrpt   10  20356729.998 ±  306557.117  ops/s
PackedIntsDecodeBenchmark.decode32ToLong           LE  thrpt   10   9907810.235 ± 1800010.522  ops/s
```

gf2121 committed:
use int

gf2121 committed:
Merge remote-tracking branch 'origin/main' into LUCENE-10315

jpountz commented on lucene/core/src/java/org/apache/lucene/util/bkd/BKDForUtil.java:
```diff
@@ -0,0 +1,88 @@
+// Encodes multiple integers in a long to get SIMD-like speedups.
+// If bitsPerValue <= 8 then we pack 8 ints per long
+// else if bitsPerValue <= 16 we pack 4 ints per long
+// else we pack 2 ints per long
```
this comment is no longer relevant

gf2121 committed:
spotless

jpountz commented on lucene/core/src/java/org/apache/lucene/util/bkd/DocIdsWriter.java:
```diff
@@ -24,58 +24,94 @@
+  private static final byte BPV_24 = (byte) 24;
+  private static final byte BPV_24_FOR_UTIL = (byte) 32 + 24;
+  private static final byte BPV_32 = (byte) 32;
+  private static final byte BPV_32_FOR_UTIL = (byte) 32 + 32;
```
do BPV_32 and BPV_32_FOR_UTIL encode data the same way? If yes, could we combine them?

jpountz commented on lucene/core/src/java/org/apache/lucene/util/bkd/DocIdsWriter.java:
```diff
@@ -24,58 +24,94 @@
+  private static final byte DELTA_VINT = (byte) 0;
+  private static final byte DELTA_FOR_UTIL = (byte) 32 + 16;
+  private static final byte BPV_24 = (byte) 24;
+  private static final byte BPV_24_FOR_UTIL = (byte) 32 + 24;
```
maybe call them LEGACY_BPV_24 and BPV_24 instead to make it clear which one is the right one

jpountz reviewed (commented):
I'd be curious to get @iverase take on this PR since he opened a PR for a similar change a while back.

It would be nice to fork this into a new codec which wouldn't have to support e.g. the vint encoding any longer, though I know it's not easy with BKDWriter/BKDReader. Maybe we should at least mark the write logic that should no longer be used as deprecated to avoid using it mistakenly.

gf2121 committed:
iter on review advice

iverase commented:
I don't like that the optimisation only works for a specific number of points, it feels very tricky. If we are doing something like that we should remove the possibility of users changing the number of points per leaf as it makes no sense any longer to change it? e.g changing it will never give you more performance.

I remember in the approach I tried I was batching the docIds by a number (128 or 256) but in general tI didn't see much better performance comparing to the added complexity.

gf2121 committed:
make bkd foruti flexible

gf2121 commented:
Thanks @iverase !

> I remember in the approach I tried I was batching the docIds by a number (128 or 256) but in general tI didn't see much better performance comparing to the added complexity.

I saw the benchmark result in https://github.com/apache/lucene-solr/pull/1538. It is similar to the benchmark result mentioned above: around 20% QPS increase for 1D high cardinality cases. But it seems like the benchmark did not cover all use cases. Try to think about some worst cases, e.g. low cardinality fields (It would be great if you can take a look at [here](https://github.com/apache/lucene/pull/541#issue-1080062490)). I proposed a BitSet approach several days ago to solve this but that's not enough since the optimization can hardly be triggered when cardinality comes to 32+.

> I don't like that the optimisation only works for a specific number of points, it feels trappy.

+1, I made a [change](https://github.com/apache/lucene/pull/541/commits/ccdae2f6920abff52abf6aae42d65b33861d4cf6) to make the BKDForutil flexible and benchmark seems the performance doesn't show a significant regression.

iverase commented:
This is looking much better. I tried some of my benchmarks and for points it looks like:

```
|Approach||Index time (sec)||Force merge time (sec)||Index size (GB)||Reader heap (MB)||
     ||Dev||Base||Diff ||Dev ||Base ||diff  ||Dev||Base||Diff||Dev||Base||Diff ||
|points|104.2s|104.2s|-0%|0.0s|0.0s| 0%|0.50|0.50|-0%|0.00|0.00| 0%|
|geo3d|135.0s|135.1s|-0%|0.0s|0.0s| 0%|0.70|0.70|-0%|0.00|0.00| 0%|
|shapes|168.6s|167.6s| 1%|0.0s|0.0s| 0%|1.26|1.26|-0%|0.00|0.00| 0%|
()
||Approach||Shape||M hits/sec   ||QPS      ||Hit count   ||
         ||Dev||Base ||Diff||Dev||Base||Diff||Dev||Base||Diff||
|points|polyRussia|16.15|16.00| 1%|4.60|4.56| 1%|3508846|3508846| 0%|
|points|polyMedium|9.33|9.33|-0%|114.28|114.33|-0%|2693559|2693559| 0%|
|points|poly 10|87.86|86.64| 1%|55.56|54.79| 1%|355809475|355809475| 0%|
|points|box|91.51|91.54|-0%|93.11|93.15|-0%|221118844|221118844| 0%|
|points|distance|86.66|85.71| 1%|50.91|50.36| 1%|382961957|382961957| 0%|
|points|nearest 10|0.03|0.03|-1%|2752.04|2783.10|-1%|60844404|60844404| 0%|
|geo3d|polyRussia|0.71|0.69| 3%|0.20|0.20| 3%|3508671|3508671| 0%|
|geo3d|polyMedium|0.74|0.74|-0%|9.10|9.10|-0%|2693541|2693541| 0%|
|geo3d|poly 10|57.57|56.69| 2%|36.40|35.84| 2%|355855227|355855227| 0%|
|geo3d|box|62.51|60.90| 3%|63.60|61.97| 3%|221118844|221118844| 0%|
|geo3d|distance|76.84|76.16| 1%|45.10|44.70| 1%|383371904|383371904| 0%|
|shapes|polyRussia|10.08|10.49|-4%|2.87|2.99|-4%|3508846|3508846| 0%|
|shapes|polyMedium|3.85|3.80| 1%|47.18|46.52| 1%|2693559|2693559| 0%|
|shapes|poly 10|44.75|41.90| 7%|28.30|26.49| 7%|355809475|355809475| 0%|
|shapes|box|47.84|44.50| 8%|48.68|45.29| 8%|221118844|221118844| 0%|
|shapes|distance|49.17|46.19| 6%|28.89|27.14| 6%|382961957|382961957| 0%|
```

There is a nice improvement for high dimensional cases. This is similar to what I was seeing except in my approach I noticed a bigger slow down for the 2-D case. I think this type of change speeds up the case when we visit the leaf node but it can slow down the case where we visit only doc_Ids. Your approach seems to limit the second case.

On the other hand, I tried another benchmark with LatLonShape and I am seeing an error. I haven't been able to dug it properly but I think we need to understand the error before moving forward:

```
 0.4s: intersects, polyRussia: Exception in thread "main" java.lang.ArrayIndexOutOfBoundsException: Index 153844 out of bounds for length 92548
    0.4s: intersects, polyRussia:       at org.apache.lucene.util.FixedBitSet.get(FixedBitSet.java:186)
    0.4s: intersects, polyRussia:       at org.apache.lucene.document.SpatialQuery$4.visit(SpatialQuery.java:479)
    0.4s: intersects, polyRussia:       at org.apache.lucene.util.bkd.BKDReader$BKDPointTree.visitCompressedDocValues(BKDReader.java:870)
    0.4s: intersects, polyRussia:       at org.apache.lucene.util.bkd.BKDReader$BKDPointTree.visitDocValuesWithCardinality(BKDReader.java:784)
    0.4s: intersects, polyRussia:       at org.apache.lucene.util.bkd.BKDReader$BKDPointTree.visitDocValues(BKDReader.java:554)
    0.4s: intersects, polyRussia:       at org.apache.lucene.util.bkd.BKDReader$BKDPointTree.visitLeavesOneByOne(BKDReader.java:539)
    0.4s: intersects, polyRussia:       at org.apache.lucene.util.bkd.BKDReader$BKDPointTree.visitDocValues(BKDReader.java:533)
    0.4s: intersects, polyRussia:       at org.apache.lucene.index.PointValues.intersect(PointValues.java:364)
    0.4s: intersects, polyRussia:       at org.apache.lucene.index.PointValues.intersect(PointValues.java:357)

```

gf2121 commented:
Thanks @iverase for the benchmark!

> On the other hand, I tried another benchmark with LatLonShape and I am seeing an error. I haven't been able to dug it properly but I think we need to understand the error before moving forward

It seems there is something wrong with the ids codec, but i can not repoduce it even by repeating test cases the whole night. I also tried to run the `IndexAndSearchOpenStreetMaps` in luceneutil, failed to reproduce too. Could you tell me how to reproduce this error locally? Thanks!

iverase commented:
You need to try `IndexAndSearchShapes` in luceneutil. Not sure if it is up-to-date but running it is similar to `IndexAndSearchOpenStreetMaps`.

gf2121 committed:
resolve conflict

gf2121 commented:
Hi @iverase ! Sorry to disturb again, but I can not see the error with the `IndexAndSearchShapes` in luceneutil too. (I run the script with param `-polyRussia -intersects -reindex` )

Could you tell me what the param you were using and post the newest script code here? Thanks a lot!

iverase commented:
have you used the data in here: http://home.apache.org/~ivera/osmdata.wkt.gz?

gf2121 commented:
@iverase Yes i've put the file under `DATA_LOCATION`.

```
➜  points ls -lh
total 10971488
-rw-r--r--@ 1 gf  staff    23M 12 15 13:36 cleveland.poly.txt.gz
-rw-r--r--  1 gf  staff   1.9G 12 15 13:42 latlon.subsetPlusAllLondon.txt
-rw-r--r--@ 1 gf  staff   938K 12 15 13:36 london.boroughs.poly.txt.gz
-rw-r--r--  1 gf  staff   3.3G  1 21 00:36 osmdata.wkt
-rw-r--r--@ 1 gf  staff    62K 12 15 13:36 russia.poly.txt.gz
```

iverase commented:
I tried to reproduce the issue in another machine again and I was able to do. In addition I was able to run it as well successfully, the difference between the two runs is that in the successful one I got this warning message:

```
WARNING: Unmapping is not supported, because Lucene cannot read 'jdk.unsupported' module [please add 'jdk.unsupported' to modular application either by command line or its module descriptor]
```

So you need to make sure you can run MMAP which points to the readInts implementation in `ByteBufferIndexInput`. Just a quick check by removing the implementation shows that the error disappears for MMAP.


I had a look into the method and I could not spot anything doggy. More over, I replaced the line:

```
guard.getInts(curIntBufferViews[position & 0x03].position(position >>> 2), dst, offset, length);
```

with: (Not reusing the int buffers)

```
IntBuffer intBuffer = curBuf.duplicate().order(ByteOrder.LITTLE_ENDIAN).position(position).asIntBuffer();
guard.getInts(intBuffer, dst, offset, length);
```

And the error disappear. Not sure what is going on, maybe a jdk bug?

gf2121 committed:
unset int buffer

gf2121 committed:
add some tests for read Ints

gf2121 committed:
spotless

gf2121 committed:
iter

gf2121 commented:
@iverase Thank you very much for helping debug so much!

> I had a look into the method and I could not spot anything strange. More over, I replaced the line:
> guard.getInts(curIntBufferViews[position & 0x03].position(position >>> 2), dst, offset, length);
with: (Not reusing the int buffers)
> IntBuffer intBuffer = curBuf.duplicate().order(ByteOrder.LITTLE_ENDIAN).position(position).asIntBuffer();
guard.getInts(intBuffer, dst, offset, length);

I found out the bug based on this clue you provided,  and fixed it in this [commit](https://github.com/apache/lucene/pull/541/commits/6a84978f32a73336db071f8a830f1f7135dce800). Also added some more tests around the `readInts`.

Very sorry for my negligence, and thanks again for you help!

iverase commented on lucene/core/src/java/org/apache/lucene/store/ByteBufferIndexInput.java:
```diff
@@ -503,6 +538,8 @@ private void unsetBuffers() {
     curBufIndex = 0;
     curLongBufferViews = null;
+    curFloatBufferViews = null;
+    curIntBufferViews = null;
```
`curFloatBufferViews` does not belong to this PR. I wonder if we should open a separate issue for this as it might lead to unknown bugs? what do you think @jpountz

iverase commented on lucene/core/src/java/org/apache/lucene/util/bkd/BKDForUtil.java:
```diff
@@ -0,0 +1,105 @@
+  private final int[] tmp;
+
+  BKDForUtil(int maxPointsInLeaf) {
+    tmp = new int[maxPointsInLeaf / 4 * 3];
```
I am a bit confused with this, can you add an explanation of why we are choosing that size. Probably it would be good to use parenthesis to clarify the order of execution,

iverase commented on lucene/core/src/java/org/apache/lucene/util/bkd/BKDForUtil.java:
```diff
@@ -0,0 +1,105 @@
+
+final class BKDForUtil {
+
+  static final int BLOCK_SIZE = 512;
```
I think this is not needed any more, should be remove it?

iverase reviewed (commented):
I think it would be nice to add more test to readInts. Probably enough to add the readInts counterpart that are on `BaseDirectoryTestCase` and `BaseChunkedDirectoryTestCase`.

Otherwise just left a few comments but I think this a good change.

gf2121 committed:
fix tests and add some notes for tmp length

gf2121 committed:
fix typo

gf2121 committed:
spotless

iverase reviewed (approved):
LGTM

gf2121 commented:
Hi @jpountz !  I wonder if you'd like to take another look at this PR? I'll merge and backport this if you agree.

jpountz commented on lucene/core/src/java/org/apache/lucene/util/bkd/BKDForUtil.java:
```diff
@@ -0,0 +1,108 @@
+    }
+    for (int i = 0; i < quarterLen; i++) {
+      final int longIdx = off + i + quarterLen3;
+      tmp[i] |= (ints[longIdx] >>> 16) & 0xFF;
```
I think we don't even need toe mask since values use 24 bits?

```suggestion
      tmp[i] |= ints[longIdx] >>> 16;
```

jpountz commented on lucene/core/src/java/org/apache/lucene/util/bkd/DocIdsWriter.java:
```diff
@@ -24,89 +24,73 @@
-        out.writeVInt(doc - previous);
-        previous = doc;
+
+    if (Integer.toUnsignedLong(min2max) <= 0xFFFFL) {
```
since doc IDs are integers in [0, MAX_VALUE) we don't need to convert to an unsigned long (MAX_VALUE is not a legal doc ID)
```suggestion
    if (min2max <= 0xFFFF) {
```

jpountz commented on lucene/core/src/java/org/apache/lucene/util/bkd/DocIdsWriter.java:
```diff
@@ -24,89 +24,73 @@
-          out.writeShort((short) (docIds[start + i] >>> 8));
-          out.writeByte((byte) docIds[start + i]);
-        }
+      if (Integer.toUnsignedLong(max) <= 0xFFFFFFL) {
```
```suggestion
      if (max <= 0xFFFFFF) {
```

jpountz reviewed (approved):
LGTM

We should look into improving testing in a follow-up, there is now little testing for the legacy vint logic.

gf2121 committed:
iter on feed back

jpountz commented on lucene/core/src/java/org/apache/lucene/util/bkd/DocIdsWriter.java:
```diff
@@ -24,89 +24,73 @@
-        out.writeVInt(doc - previous);
-        previous = doc;
+
+    if (min2max <= 0xFFFFL) {
```
```suggestion
    if (min2max <= 0xFFFF) {
```

jpountz commented on lucene/core/src/java/org/apache/lucene/util/bkd/DocIdsWriter.java:
```diff
@@ -24,89 +24,73 @@
-          out.writeShort((short) (docIds[start + i] >>> 8));
-          out.writeByte((byte) docIds[start + i]);
-        }
+      if (max <= 0xFFFFFFL) {
```
```suggestion
      if (max <= 0xFFFFFF) {
```

gf2121 committed:
Update lucene/core/src/java/org/apache/lucene/util/bkd/DocIdsWriter.java

Co-authored-by: Adrien Grand <jpountz@gmail.com>

gf2121 committed:
Update lucene/core/src/java/org/apache/lucene/util/bkd/DocIdsWriter.java

Co-authored-by: Adrien Grand <jpountz@gmail.com>

gf2121 committed:
Merge remote-tracking branch 'origin/main' into LUCENE-10315
